scipy/optimize/_qap.py

```diff
@@ -0,0 +1,789 @@
+        # generate a nxn matrix where each entry is a random number [0, 1]
+        # would use rand, but Generators don't have it
+        # would use random, but old mtrand.RandomStates don't have it
+        K = rng.uniform(size=(n_unseed, n_unseed))
+        # Sinkhorn balancing
+        K = _doubly_stochastic(K)
+        # initialize J, a doubly stochastic barycenter
+        P = J * init_weight + (1 - init_weight) * K
+    else:
+        P = J
+    const_sum = A21 @ B21.T + A12.T @ B12
+
+    # [1] Algorithm 1 Step 2 - loop while stopping criteria not met
+    for n_iter in range(1, maxiter+1):
+        # [1] Algorithm 1 Step 3 - compute the gradient of f(P) = -tr(APB^tP^t)
+        grad_fp = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+        # [1] Algorithm 1 Step 4 - get direction Q by solving Eq. 8
+        rows, cols = linear_sum_assignment(obj_func_scalar * grad_fp)
+        Q = np.zeros((n_unseed, n_unseed))
+        Q[rows, cols] = 1  # initialize search direction matrix Q
+
+
+        # [1] Algorithm 1 Step 5 - compute the step size
+        # Noting that e.g. trace(Ax) = trace(A)*x, expand and re-collect
+        # terms as ax**2 + bx + c. c does not affect location of minimum
+        # and can be ignored. Also, note that trace(A@B) = (A.T*B).sum();
+        # apply where possible for efficiency.
+        R = P - Q
+        b21 = ((R.T @ A21) * B21).sum()
+        b12 = ((R.T @ A12.T) * B12.T).sum()
+        AR22 = A22.T @ R
+        BR22 = B22 @ R.T
+        b22a = (AR22.T * (B22 @ Q.T)).sum()
+        b22b = ((A22.T @ Q).T * BR22).sum()
+        a = (AR22.T * BR22).sum()
+        b = b21 + b12 + b22a + b22b
+        # critical point of ax^2 + bx + c is at x = -d/(2*e)
+        # if a * obj_func_scalar > 0, it is a minimum
+        # if minimum is not in [0, 1], only endpoints need to be considered
+        if a*obj_func_scalar > 0 and 0 <= -b/(2*a) <= 1:
+            alpha = -b/(2*a)
+        else:
+            alpha = np.argmin([0, (b + a)*obj_func_scalar])
+
+        # [1] Algorithm 1 Step 6 - Update P
+        P_i1 = alpha * P + (1 - alpha) * Q
+        if np.linalg.norm(P - P_i1) < tol:
+            P = P_i1
+            break
+        P = P_i1
+    # [1] Algorithm 1 Step 7 - end main loop
+
+    # [1] Algorithm 1 Step 8 - project onto the set of permutation matrices
+    _, col = linear_sum_assignment(-P)
+    perm = np.concatenate((np.arange(n_seeds), col + n_seeds))
+
+    unshuffled_perm = np.zeros(n, dtype=int)
+    unshuffled_perm[perm_A] = perm_B[perm]
+
+    score = _calc_score(A, B, unshuffled_perm)
+
+    res = {"col_ind": unshuffled_perm, "fun": score, "nit": n_iter}
+
+    return OptimizeResult(res)
+
+
+def _check_init_input(init_J, n):
+    row_sum = np.sum(init_J, axis=0)
+    col_sum = np.sum(init_J, axis=1)
+    tol = 1e-3
+    msg = None
+    if init_J.shape != (n, n):
+        msg = "`init_J` matrix must have same shape as A and B"
+    elif ((~np.isclose(row_sum, np.ones(n), atol=tol)).any() or
```

Comment:
You can just use `1` instead of `np.ones(n)`

Reply:
Done